File: android/src/main/kotlin/com/example/flutter_storage_manager/FlutterStorageManagerPlugin.kt
package com.example.flutter_storage_manager

import android.content.Context
import android.os.Build
import android.os.Environment
import android.os.StatFs
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URL

class FlutterStorageManagerPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel : MethodChannel
  private lateinit var context: Context
  private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
  private val downloadJobs = mutableMapOf<String, Job>()

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_storage_manager")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.applicationContext
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "getStorageInfo" -> scope.launch {
        try {
          val info = getStorageInfo()
          withContext(Dispatchers.Main) { result.success(info) }
        } catch (e: Exception) {
          withContext(Dispatchers.Main) { result.error("unknown", e.message, null) }
        }
      }
      "analyzeStorage" -> scope.launch {
        try {
          val analysis = analyzeStorage()
          withContext(Dispatchers.Main) { result.success(analysis) }
        } catch (e: Exception) {
          withContext(Dispatchers.Main) { result.error("unknown", e.message, null) }
        }
      }
      "getItems" -> scope.launch {
        val categoryStr = call.argument<String>("category")
        try {
          val items = getItems(categoryStr)
          withContext(Dispatchers.Main) { result.success(items) }
        } catch (e: Exception) {
          withContext(Dispatchers.Main) { result.error("unknown", e.message, null) }
        }
      }
      "deleteItems" -> scope.launch {
        val paths = call.argument<List<String>>("paths") ?: emptyList()
        try {
          val res = deleteItems(paths)
          withContext(Dispatchers.Main) { result.success(res) }
        } catch (e: Exception) {
          withContext(Dispatchers.Main) { result.error("unknown", e.message, null) }
        }
      }
      "clearCategory" -> scope.launch {
        val categoryStr = call.argument<String>("category")
        try {
          val res = clearCategory(categoryStr)
          withContext(Dispatchers.Main) { result.success(res) }
        } catch (e: Exception) {
          withContext(Dispatchers.Main) { result.error("unknown", e.message, null) }
        }
      }
      "clearAllManagedFiles" -> scope.launch {
        try {
          val res = clearAllManagedFiles()
          withContext(Dispatchers.Main) { result.success(res) }
        } catch (e: Exception) {
          withContext(Dispatchers.Main) { result.error("unknown", e.message, null) }
        }
      }
      "fileExists" -> scope.launch {
        val path = call.argument<String>("path") ?: ""
        try {
          val exists = fileExists(path)
          withContext(Dispatchers.Main) { result.success(exists) }
        } catch (e: Exception) {
          withContext(Dispatchers.Main) { result.error("unknown", e.message, null) }
        }
      }
      "getFileInfo" -> scope.launch {
        val path = call.argument<String>("path") ?: ""
        try {
          val info = getFileInfo(path)
          withContext(Dispatchers.Main) { result.success(info) }
        } catch (e: Exception) {
          withContext(Dispatchers.Main) { result.error("unknown", e.message, null) }
        }
      }
      "prefetchFile" -> scope.launch {
        val urlStr = call.argument<String>("url") ?: ""
        val fileName = call.argument<String>("fileName")
        val categoryStr = call.argument<String>("category") ?: "prefetched"
        val overwrite = call.argument<Boolean>("overwrite") ?: false
        try {
          prefetchFile(urlStr, fileName, categoryStr, overwrite)
          withContext(Dispatchers.Main) { result.success(null) }
        } catch (e: Exception) {
          withContext(Dispatchers.Main) { result.error("networkError", e.message, null) }
        }
      }
      "canDelete" -> scope.launch {
        val path = call.argument<String>("path") ?: ""
        try {
          val can = isSafePath(File(path))
          withContext(Dispatchers.Main) { result.success(can) }
        } catch (e: Exception) {
          withContext(Dispatchers.Main) { result.error("unknown", e.message, null) }
        }
      }
      "checkPermission", "requestPermission" -> {
        // We only use internal app dirs, so permission is not required on any modern Android.
        result.success("notRequired")
      }
      "getCapabilities" -> {
        result.success(mapOf(
          "canAnalyze" to true,
          "canClearCache" to true,
          "canClearTemporary" to true,
          "canPrefetch" to true,
          "canAccessUserSelectedFiles" to false,
          "canShowExactPaths" to true
        ))
      }
      else -> result.notImplemented()
    }
  }

  private fun getCategoryDir(category: String): File {
    return when (category) {
      "cache" -> context.cacheDir
      "temporary" -> File(context.cacheDir, "temporary").apply { mkdirs() }
      "prefetched" -> File(context.filesDir, "prefetched").apply { mkdirs() }
      "downloads" -> File(context.filesDir, "downloads").apply { mkdirs() }
      "generated" -> File(context.filesDir, "generated").apply { mkdirs() }
      "thumbnails" -> File(context.cacheDir, "thumbnails").apply { mkdirs() }
      "pluginManaged" -> File(context.filesDir, "pluginManaged").apply { mkdirs() }
      else -> File(context.filesDir, "pluginManaged").apply { mkdirs() }
    }
  }

  private fun getAllowedRoots(): List<String> {
    return listOf(
      context.cacheDir.canonicalPath,
      context.filesDir.canonicalPath
    )
  }

  private fun isSafePath(file: File): Boolean {
    val canonicalPath = file.canonicalPath
    for (root in getAllowedRoots()) {
      if (canonicalPath.startsWith(root)) {
        return true
      }
    }
    return false
  }

  private fun fileToMap(file: File, category: String): Map<String, Any?> {
    return mapOf(
      "id" to file.canonicalPath,
      "name" to file.name,
      "path" to file.canonicalPath,
      "sizeBytes" to (if (file.isDirectory) getFolderSize(file) else file.length()),
      "category" to category,
      "createdAt" to file.lastModified(), // Android doesn't expose creation time easily
      "modifiedAt" to file.lastModified(),
      "canDelete" to true,
      "isDirectory" to file.isDirectory
    )
  }

  private fun getFolderSize(folder: File): Long {
    var length: Long = 0
    folder.listFiles()?.forEach { file ->
      length += if (file.isDirectory) getFolderSize(file) else file.length()
    }
    return length
  }

  private fun getStorageInfo(): Map<String, Any?> {
    val stat = StatFs(context.filesDir.path)
    val blockSize = stat.blockSizeLong
    val totalBlocks = stat.blockCountLong
    val availableBlocks = stat.availableBlocksLong

    val totalBytes = totalBlocks * blockSize
    val freeBytes = availableBlocks * blockSize
    val usedBytes = totalBytes - freeBytes

    val appUsedBytes = getFolderSize(context.filesDir) + getFolderSize(context.cacheDir)
    val appCacheBytes = getFolderSize(context.cacheDir)

    return mapOf(
      "totalBytes" to totalBytes,
      "usedBytes" to usedBytes,
      "freeBytes" to freeBytes,
      "appUsedBytes" to appUsedBytes,
      "appCacheBytes" to appCacheBytes,
      "usagePercentage" to if (totalBytes > 0) (usedBytes.toDouble() / totalBytes.toDouble()) * 100.0 else 0.0
    )
  }

  private fun analyzeStorage(): Map<String, Any?> {
    val categories = listOf("cache", "temporary", "prefetched", "downloads", "generated", "thumbnails", "pluginManaged")
    val catInfos = mutableListOf<Map<String, Any>>()
    val allItems = mutableListOf<Map<String, Any?>>()
    var totalCleanable = 0L

    for (cat in categories) {
      val dir = getCategoryDir(cat)
      var size = 0L
      var count = 0
      if (dir.exists()) {
        dir.listFiles()?.forEach { file ->
          if (file.name != "temporary" && file.name != "thumbnails") { // don't double count if inside cacheDir
             val fSize = if (file.isDirectory) getFolderSize(file) else file.length()
             size += fSize
             count += 1
             allItems.add(fileToMap(file, cat))
          }
        }
      }
      totalCleanable += size
      catInfos.add(mapOf(
        "category" to cat,
        "sizeBytes" to size,
        "itemCount" to count
      ))
    }

    return mapOf(
      "categories" to catInfos,
      "totalCleanableBytes" to totalCleanable,
      "items" to allItems
    )
  }

  private fun getItems(categoryStr: String?): List<Map<String, Any?>> {
    val items = mutableListOf<Map<String, Any?>>()
    if (categoryStr != null) {
      val dir = getCategoryDir(categoryStr)
      dir.listFiles()?.forEach { file ->
        items.add(fileToMap(file, categoryStr))
      }
    } else {
      val categories = listOf("cache", "temporary", "prefetched", "downloads", "generated", "thumbnails", "pluginManaged")
      for (cat in categories) {
        val dir = getCategoryDir(cat)
        dir.listFiles()?.forEach { file ->
           items.add(fileToMap(file, cat))
        }
      }
    }
    return items
  }

  private fun deleteItems(paths: List<String>): Map<String, Any?> {
    var deletedCount = 0
    var failedCount = 0
    var skippedCount = 0
    var bytesFreed = 0L
    val results = mutableListOf<Map<String, Any?>>()

    for (path in paths) {
      val file = File(path)
      if (!file.exists()) {
        skippedCount++
        results.add(mapOf("path" to path, "success" to false, "bytes" to 0, "errorCode" to "fileNotFound", "message" to "File does not exist"))
        continue
      }
      if (!isSafePath(file)) {
        skippedCount++
        results.add(mapOf("path" to path, "success" to false, "bytes" to 0, "errorCode" to "accessDenied", "message" to "Path is outside allowed directories"))
        continue
      }
      val size = if (file.isDirectory) getFolderSize(file) else file.length()
      val success = file.deleteRecursively()
      if (success) {
        deletedCount++
        bytesFreed += size
        results.add(mapOf("path" to path, "success" to true, "bytes" to size))
      } else {
        failedCount++
        results.add(mapOf("path" to path, "success" to false, "bytes" to 0, "errorCode" to "ioError", "message" to "Failed to delete file"))
      }
    }
    return mapOf(
      "deletedCount" to deletedCount,
      "failedCount" to failedCount,
      "skippedCount" to skippedCount,
      "bytesFreed" to bytesFreed,
      "results" to results
    )
  }

  private fun clearCategory(categoryStr: String?): Map<String, Any?> {
    if (categoryStr == null) return mapOf()
    val dir = getCategoryDir(categoryStr)
    val paths = dir.listFiles()?.map { it.absolutePath } ?: emptyList()
    return deleteItems(paths)
  }

  private fun clearAllManagedFiles(): Map<String, Any?> {
    val categories = listOf("cache", "temporary", "prefetched", "downloads", "generated", "thumbnails", "pluginManaged")
    val paths = mutableListOf<String>()
    for (cat in categories) {
      val dir = getCategoryDir(cat)
      dir.listFiles()?.forEach { file ->
        if (file.name != "temporary" && file.name != "thumbnails") { // Prevent deleting sub-category folders by mistake if scanning root
          paths.add(file.absolutePath)
        }
      }
    }
    return deleteItems(paths)
  }

  private fun fileExists(path: String): Boolean {
    val file = File(path)
    return file.exists() && isSafePath(file)
  }

  private fun getFileInfo(path: String): Map<String, Any?>? {
    val file = File(path)
    if (!file.exists() || !isSafePath(file)) return null
    return fileToMap(file, "pluginManaged") // Unknown cat
  }

  private fun prefetchFile(urlStr: String, fileName: String?, categoryStr: String, overwrite: Boolean) {
    val url = URL(urlStr)
    val dir = getCategoryDir(categoryStr)
    val safeFileName = fileName ?: urlStr.substringAfterLast("/").takeIf { it.isNotEmpty() } ?: "downloaded_file"
    val file = File(dir, safeFileName)
    
    if (file.exists() && !overwrite) {
      return
    }

    val connection = url.openConnection() as HttpURLConnection
    connection.requestMethod = "GET"
    connection.connectTimeout = 15000
    connection.readTimeout = 15000
    connection.connect()

    if (connection.responseCode != HttpURLConnection.HTTP_OK) {
      throw Exception("HTTP \${connection.responseCode}")
    }

    val input = connection.inputStream
    val output = FileOutputStream(file)

    try {
      val buffer = ByteArray(4096)
      var bytesRead: Int
      while (input.read(buffer).also { bytesRead = it } != -1) {
        output.write(buffer, 0, bytesRead)
      }
    } finally {
      output.close()
      input.close()
      connection.disconnect()
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    scope.cancel()
  }
}

File: ios/Classes/FlutterStorageManagerPlugin.swift
import Flutter
import UIKit

public class FlutterStorageManagerPlugin: NSObject, FlutterPlugin {
  private let queue = DispatchQueue(label: "com.example.flutter_storage_manager.queue", attributes: .concurrent)

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_storage_manager", binaryMessenger: registrar.messenger())
    let instance = FlutterStorageManagerPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    queue.async {
      do {
        switch call.method {
        case "getStorageInfo":
          let info = try self.getStorageInfo()
          DispatchQueue.main.async { result(info) }
        case "analyzeStorage":
          let analysis = try self.analyzeStorage()
          DispatchQueue.main.async { result(analysis) }
        case "getItems":
          let args = call.arguments as? [String: Any]
          let category = args?["category"] as? String
          let items = try self.getItems(categoryStr: category)
          DispatchQueue.main.async { result(items) }
        case "deleteItems":
          let args = call.arguments as? [String: Any]
          let paths = args?["paths"] as? [String] ?? []
          let res = try self.deleteItems(paths: paths)
          DispatchQueue.main.async { result(res) }
        case "clearCategory":
          let args = call.arguments as? [String: Any]
          let category = args?["category"] as? String
          let res = try self.clearCategory(categoryStr: category)
          DispatchQueue.main.async { result(res) }
        case "clearAllManagedFiles":
          let res = try self.clearAllManagedFiles()
          DispatchQueue.main.async { result(res) }
        case "fileExists":
          let args = call.arguments as? [String: Any]
          let path = args?["path"] as? String ?? ""
          let exists = self.fileExists(path: path)
          DispatchQueue.main.async { result(exists) }
        case "getFileInfo":
          let args = call.arguments as? [String: Any]
          let path = args?["path"] as? String ?? ""
          let info = self.getFileInfo(path: path)
          DispatchQueue.main.async { result(info) }
        case "prefetchFile":
          let args = call.arguments as? [String: Any]
          let url = args?["url"] as? String ?? ""
          let fileName = args?["fileName"] as? String
          let categoryStr = args?["category"] as? String ?? "prefetched"
          let overwrite = args?["overwrite"] as? Bool ?? false
          try self.prefetchFile(urlStr: url, fileName: fileName, categoryStr: categoryStr, overwrite: overwrite)
          DispatchQueue.main.async { result(nil) }
        case "canDelete":
          let args = call.arguments as? [String: Any]
          let path = args?["path"] as? String ?? ""
          let can = self.isSafePath(path: path)
          DispatchQueue.main.async { result(can) }
        case "checkPermission", "requestPermission":
          DispatchQueue.main.async { result("notRequired") }
        case "getCapabilities":
          let caps: [String: Any] = [
            "canAnalyze": true,
            "canClearCache": true,
            "canClearTemporary": true,
            "canPrefetch": true,
            "canAccessUserSelectedFiles": false,
            "canShowExactPaths": true
          ]
          DispatchQueue.main.async { result(caps) }
        default:
          DispatchQueue.main.async { result(FlutterMethodNotImplemented) }
        }
      } catch {
        DispatchQueue.main.async {
          result(FlutterError(code: "unknown", message: error.localizedDescription, details: nil))
        }
      }
    }
  }

  private func getCategoryUrl(_ category: String) throws -> URL {
    let fm = FileManager.default
    let caches = try fm.url(for: .cachesDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
    let appSupport = try fm.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
    let temp = fm.temporaryDirectory

    var url: URL
    switch category {
    case "cache": url = caches
    case "temporary": url = temp.appendingPathComponent("flutter_storage_manager")
    case "prefetched": url = appSupport.appendingPathComponent("prefetched")
    case "downloads": url = appSupport.appendingPathComponent("downloads")
    case "generated": url = appSupport.appendingPathComponent("generated")
    case "thumbnails": url = caches.appendingPathComponent("thumbnails")
    case "pluginManaged": url = appSupport.appendingPathComponent("pluginManaged")
    default: url = appSupport.appendingPathComponent("pluginManaged")
    }

    if !fm.fileExists(atPath: url.path) {
      try fm.createDirectory(at: url, withIntermediateDirectories: true, attributes: nil)
    }
    return url
  }

  private func getAllowedRoots() throws -> [String] {
    let fm = FileManager.default
    let caches = try fm.url(for: .cachesDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
    let appSupport = try fm.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
    let temp = fm.temporaryDirectory

    return [
      caches.resolvingSymlinksInPath().path,
      appSupport.resolvingSymlinksInPath().path,
      temp.resolvingSymlinksInPath().path
    ]
  }

  private func isSafePath(path: String) -> Bool {
    let fm = FileManager.default
    let canonical = URL(fileURLWithPath: path).resolvingSymlinksInPath().path
    do {
      let roots = try getAllowedRoots()
      for root in roots {
        if canonical.hasPrefix(root) {
          return true
        }
      }
    } catch {
      return false
    }
    return false
  }

  private func getFolderSize(url: URL) -> Int {
    let fm = FileManager.default
    var size = 0
    if let enumerator = fm.enumerator(at: url, includingPropertiesForKeys: [.totalFileAllocatedSizeKey, .fileAllocatedSizeKey], options: []) {
      for case let fileURL as URL in enumerator {
        do {
          let resourceValues = try fileURL.resourceValues(forKeys: [.totalFileAllocatedSizeKey, .fileAllocatedSizeKey])
          size += resourceValues.totalFileAllocatedSize ?? resourceValues.fileAllocatedSize ?? 0
        } catch { }
      }
    }
    return size
  }

  private func fileToMap(url: URL, category: String) -> [String: Any]? {
    let fm = FileManager.default
    do {
      let attrs = try fm.attributesOfItem(atPath: url.path)
      let isDir = (attrs[.type] as? FileAttributeType) == .typeDirectory
      let size = isDir ? getFolderSize(url: url) : (attrs[.size] as? Int ?? 0)

      return [
        "id": url.path,
        "name": url.lastPathComponent,
        "path": url.path,
        "sizeBytes": size,
        "category": category,
        "createdAt": (attrs[.creationDate] as? Date)?.timeIntervalSince1970 ?? 0 * 1000,
        "modifiedAt": (attrs[.modificationDate] as? Date)?.timeIntervalSince1970 ?? 0 * 1000,
        "canDelete": true,
        "isDirectory": isDir
      ]
    } catch {
      return nil
    }
  }

  private func getStorageInfo() throws -> [String: Any] {
    let fm = FileManager.default
    let caches = try fm.url(for: .cachesDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
    let appSupport = try fm.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true)

    let sysAttrs = try fm.attributesOfFileSystem(forPath: caches.path)
    let totalBytes = sysAttrs[.systemSize] as? Int ?? 0
    let freeBytes = sysAttrs[.systemFreeSize] as? Int ?? 0
    let usedBytes = totalBytes - freeBytes

    let appCacheBytes = getFolderSize(url: caches)
    let appUsedBytes = appCacheBytes + getFolderSize(url: appSupport)

    return [
      "totalBytes": totalBytes,
      "usedBytes": usedBytes,
      "freeBytes": freeBytes,
      "appUsedBytes": appUsedBytes,
      "appCacheBytes": appCacheBytes,
      "usagePercentage": totalBytes > 0 ? (Double(usedBytes) / Double(totalBytes)) * 100.0 : 0.0
    ]
  }

  private func analyzeStorage() throws -> [String: Any] {
    let categories = ["cache", "temporary", "prefetched", "downloads", "generated", "thumbnails", "pluginManaged"]
    var catInfos = [[String: Any]]()
    var allItems = [[String: Any]]()
    var totalCleanable = 0
    let fm = FileManager.default

    for cat in categories {
      let url = try getCategoryUrl(cat)
      var size = 0
      var count = 0

      if let contents = try? fm.contentsOfDirectory(at: url, includingPropertiesForKeys: nil, options: .skipsHiddenFiles) {
        for fileUrl in contents {
          if fileUrl.lastPathComponent == "thumbnails" || fileUrl.lastPathComponent == "flutter_storage_manager" {
            continue
          }
          if let map = fileToMap(url: fileUrl, category: cat) {
            allItems.append(map)
            size += map["sizeBytes"] as? Int ?? 0
            count += 1
          }
        }
      }
      totalCleanable += size
      catInfos.append([
        "category": cat,
        "sizeBytes": size,
        "itemCount": count
      ])
    }

    return [
      "categories": catInfos,
      "totalCleanableBytes": totalCleanable,
      "items": allItems
    ]
  }

  private func getItems(categoryStr: String?) throws -> [[String: Any]] {
    var items = [[String: Any]]()
    let fm = FileManager.default

    if let cat = categoryStr {
      let url = try getCategoryUrl(cat)
      if let contents = try? fm.contentsOfDirectory(at: url, includingPropertiesForKeys: nil, options: .skipsHiddenFiles) {
        for fileUrl in contents {
          if let map = fileToMap(url: fileUrl, category: cat) {
            items.append(map)
          }
        }
      }
    } else {
      let categories = ["cache", "temporary", "prefetched", "downloads", "generated", "thumbnails", "pluginManaged"]
      for cat in categories {
        let url = try getCategoryUrl(cat)
        if let contents = try? fm.contentsOfDirectory(at: url, includingPropertiesForKeys: nil, options: .skipsHiddenFiles) {
          for fileUrl in contents {
            if fileUrl.lastPathComponent == "thumbnails" || fileUrl.lastPathComponent == "flutter_storage_manager" {
              continue
            }
            if let map = fileToMap(url: fileUrl, category: cat) {
              items.append(map)
            }
          }
        }
      }
    }
    return items
  }

  private func deleteItems(paths: [String]) throws -> [String: Any] {
    let fm = FileManager.default
    var deletedCount = 0
    var failedCount = 0
    var skippedCount = 0
    var bytesFreed = 0
    var results = [[String: Any]]()

    for path in paths {
      if !fm.fileExists(atPath: path) {
        skippedCount += 1
        results.append(["path": path, "success": false, "bytes": 0, "errorCode": "fileNotFound", "message": "File does not exist"])
        continue
      }
      if !isSafePath(path: path) {
        skippedCount += 1
        results.append(["path": path, "success": false, "bytes": 0, "errorCode": "accessDenied", "message": "Path is outside allowed directories"])
        continue
      }

      let url = URL(fileURLWithPath: path)
      let attrs = try? fm.attributesOfItem(atPath: path)
      let isDir = (attrs?[.type] as? FileAttributeType) == .typeDirectory
      let size = isDir ? getFolderSize(url: url) : (attrs?[.size] as? Int ?? 0)

      do {
        try fm.removeItem(at: url)
        deletedCount += 1
        bytesFreed += size
        results.append(["path": path, "success": true, "bytes": size])
      } catch {
        failedCount += 1
        results.append(["path": path, "success": false, "bytes": 0, "errorCode": "ioError", "message": error.localizedDescription])
      }
    }

    return [
      "deletedCount": deletedCount,
      "failedCount": failedCount,
      "skippedCount": skippedCount,
      "bytesFreed": bytesFreed,
      "results": results
    ]
  }

  private func clearCategory(categoryStr: String?) throws -> [String: Any] {
    guard let cat = categoryStr else { return [:] }
    let url = try getCategoryUrl(cat)
    let fm = FileManager.default
    var paths = [String]()

    if let contents = try? fm.contentsOfDirectory(at: url, includingPropertiesForKeys: nil, options: .skipsHiddenFiles) {
      for fileUrl in contents {
        paths.append(fileUrl.path)
      }
    }
    return try deleteItems(paths: paths)
  }

  private func clearAllManagedFiles() throws -> [String: Any] {
    let categories = ["cache", "temporary", "prefetched", "downloads", "generated", "thumbnails", "pluginManaged"]
    var paths = [String]()
    let fm = FileManager.default

    for cat in categories {
      let url = try getCategoryUrl(cat)
      if let contents = try? fm.contentsOfDirectory(at: url, includingPropertiesForKeys: nil, options: .skipsHiddenFiles) {
        for fileUrl in contents {
          if fileUrl.lastPathComponent == "thumbnails" || fileUrl.lastPathComponent == "flutter_storage_manager" {
            continue
          }
          paths.append(fileUrl.path)
        }
      }
    }
    return try deleteItems(paths: paths)
  }

  private func fileExists(path: String) -> Bool {
    return FileManager.default.fileExists(atPath: path) && isSafePath(path: path)
  }

  private func getFileInfo(path: String) -> [String: Any]? {
    if !fileExists(path: path) { return nil }
    return fileToMap(url: URL(fileURLWithPath: path), category: "pluginManaged")
  }

  private func prefetchFile(urlStr: String, fileName: String?, categoryStr: String, overwrite: Bool) throws {
    guard let url = URL(string: urlStr) else { throw NSError(domain: "", code: -1, userInfo: nil) }
    let dir = try getCategoryUrl(categoryStr)
    let safeFileName = fileName ?? url.lastPathComponent
    let dest = dir.appendingPathComponent(safeFileName == "" ? "downloaded_file" : safeFileName)

    let fm = FileManager.default
    if fm.fileExists(atPath: dest.path) && !overwrite {
      return
    }

    let semaphore = DispatchSemaphore(value: 0)
    var downloadError: Error?

    let task = URLSession.shared.downloadTask(with: url) { tempLocalUrl, response, error in
      if let error = error {
        downloadError = error
      } else if let tempLocalUrl = tempLocalUrl {
        do {
          if fm.fileExists(atPath: dest.path) {
            try fm.removeItem(at: dest)
          }
          try fm.moveItem(at: tempLocalUrl, to: dest)
        } catch let e {
          downloadError = e
        }
      }
      semaphore.signal()
    }

    task.resume()
    semaphore.wait()

    if let error = downloadError {
      throw error
    }
  }
}
